bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/CheckAndMutateUtil.java

```diff
@@ -71,67 +70,129 @@ public static boolean wasMutationApplied(
+  public static class RequestBuilder {
+    private final HBaseRequestAdapter hbaseAdapter;
+    private final CheckAndMutateRowRequest.Builder requestBuilder =
+        CheckAndMutateRowRequest.newBuilder();
+
+    private final byte[] row;
+    private final byte[] family;
+    private byte[] qualifier;
+    private CompareOp compareOp;
+    private byte[] value;
+    private boolean checkNonExistence = false;
+
+    /**
+     * <p>
+     * RequestBuilder.
+     * </p>
+     * @param hbaseAdapter a {@link HBaseRequestAdapter} used to convert HBase
+     *          {@link org.apache.hadoop.hbase.client.Mutation} to Cloud Bigtable {@link Mutation}
+     * @param row the RowKey in which to to check value matching
+     * @param family the family in which to check value matching.
+     */
+    public RequestBuilder(HBaseRequestAdapter hbaseAdapter, byte[] row, byte[] family) {
+      requestBuilder.setRowKey(ByteString.copyFrom(row));
+      this.row = Preconditions.checkNotNull(row, "row is null");
+      this.family = Preconditions.checkNotNull(family, "family is null");
+
+      // TODO (issue #1709): The hbaseAdapter used here should not set client-side timestamps.
+      this.hbaseAdapter = hbaseAdapter;
+      requestBuilder.setTableName(hbaseAdapter.getBigtableTableName().toString());
     }
 
-    CheckAndMutateRowRequest.Builder requestBuilder =
-        CheckAndMutateRowRequest.newBuilder();
+    public RequestBuilder qualifier(byte[] qualifier) {
+      this.qualifier = Preconditions.checkNotNull(qualifier, "qualifier is null. Consider using" +
+          " an empty byte array, or just do not call this method if you want a null qualifier");
+      return this;
+    }
 
-    requestBuilder.setTableName(hbaseAdapter.getBigtableTableName().toString());
-
-    requestBuilder.setRowKey(ByteString.copyFrom(row));
-    Scan scan = new Scan().addColumn(family, qualifier);
-    scan.setMaxVersions(1);
-    if (value == null) {
-      // If we don't have a value and we are doing CompareOp.EQUAL, we want to mutate if there
-      // is no cell with the qualifier. If we are doing CompareOp.NOT_EQUAL, we want to mutate
-      // if there is any cell. We don't actually want an extra filter for either of these cases,
-      // but we do need to invert the compare op.
-      if (CompareFilter.CompareOp.EQUAL.equals(compareOp)) {
-        requestBuilder.addAllFalseMutations(mutations);
-      } else if (CompareFilter.CompareOp.NOT_EQUAL.equals(compareOp)) {
+    public RequestBuilder ifNotExists() {
+      Preconditions.checkState(compareOp == null,
+          "ifNotExists and ifMatches are mutually exclusive");
+      this.checkNonExistence = true;
+      return this;
+    }
+
+    public RequestBuilder ifEquals(CompareOp compareOp, byte[] value) {
+      return ifMatches(CompareOp.EQUAL, value);
+    }
+
+    public RequestBuilder ifMatches(CompareOp compareOp, byte[] value) {
```

Comment:
I think it would be cleaner to stick to a builder style and have `setOperator` & `setOperand` or go the other way make this a DSL and add the qualifier as a parameter

Reply:
HBase uses this approach, and I think I'd like to keep it for now.   I'd like to think through what a more powerful `CheckAndMutateRequest` builder would look like.  

The `with(Put|Delete|Mutation)` API allows for more than the single mutation, which is great.  It would be awesome to be able to add more conditions as well; something like:

```
builder
   .ifQualifierMatches(q1, ...)
   .ifQualifierNotExists(q2)
   .ifCondition(randomFilter)
`